mm2src/coins/solana/spl.rs

```diff
@@ -0,0 +1,457 @@
+use super::{CoinBalance, HistorySyncState, MarketCoinOps, MmCoin, SwapOps, TradeFee, TransactionEnum, TransactionFut};
+use crate::common::Future01CompatExt;
+use crate::solana::solana_common::{lamports_to_sol, ui_amount_to_amount, SufficientBalanceError};
+use crate::solana::{solana_common, AccountError, SolanaAsyncCommonOps, SolanaCommonOps, SolanaFeeDetails};
+use crate::{BalanceFut, FeeApproxStage, FoundSwapTxSpend, NegotiateSwapContractAddrErr, SolanaCoin, TradePreimageFut,
+            TradePreimageResult, TradePreimageValue, TransactionDetails, TransactionType, ValidateAddressResult,
+            ValidatePaymentInput, WithdrawError, WithdrawFut, WithdrawRequest, WithdrawResult};
+use async_trait::async_trait;
+use bigdecimal::BigDecimal;
+use bincode::serialize;
+use common::mm_error::prelude::MapToMmResult;
+use common::{mm_ctx::MmArc, mm_error::MmError, mm_number::MmNumber, now_ms};
+use futures::{FutureExt, TryFutureExt};
+use futures01::Future;
+use keys::KeyPair;
+use rpc::v1::types::Bytes as BytesJson;
+use serde_json::Value as Json;
+use solana_client::{nonblocking::rpc_client::RpcClient, rpc_request::TokenAccountsFilter};
+use solana_sdk::message::Message;
+use solana_sdk::transaction::Transaction;
+use solana_sdk::{pubkey::Pubkey, signature::Signer};
+use spl_associated_token_account::{create_associated_token_account, get_associated_token_address};
+use std::{convert::TryFrom,
+          fmt::{Debug, Formatter, Result as FmtResult},
+          str::FromStr,
+          sync::Arc};
+
+#[derive(Debug)]
+pub enum SplTokenCreationError {
+    InvalidPubkey(String),
+}
+
+#[derive(Debug)]
+pub struct SplTokenConf {
+    pub decimals: u8,
+    pub ticker: String,
+    pub token_contract_address: Pubkey,
+}
+
+#[derive(Clone, Debug)]
+pub struct SplTokenInfo {
+    pub token_contract_address: Pubkey,
+    pub decimals: u8,
+}
+
+#[derive(Debug)]
+pub struct SplProtocolConf {
+    pub platform_coin_ticker: String,
+    pub decimals: u8,
+    pub token_contract_address: String,
```

Comment:
Specific types are always preferred, this should be `Pubkey`. This will avoid returning an error from the `SplToken::new` fn https://github.com/KomodoPlatform/atomicDEX-API/blob/8fba90aef5833e06424ecdcb300399f53afb2e2d/mm2src/coins/solana/spl.rs#L64-L69
as the `InvalidPubkey` error will be returned early from the `try_from_coin_protocol` fn instead, which is preferred.
This in turn will allow you to use `map` here https://github.com/KomodoPlatform/atomicDEX-API/pull/1109#discussion_r823465549

Reply:
It's was a string but I was not able to avoid Solana sdk crate in coins_activation then

Reply:
You can use the `TryInto` trait to get the `Pubkey` from the `String` without adding Solana SDK as a dependency in coins_activation

Reply:
Will try that shortly